registry/storage/driver/s3-aws/s3.go

```diff
@@ -1070,7 +1070,33 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
 	}
 
 	var objectCount int64
-	if err := d.doWalk(ctx, &objectCount, d.s3Path(path), prefix, f); err != nil {
+	if err := d.doWalk(ctx, &objectCount, d.s3Path(path), prefix, true, f); err != nil {
+		return err
+	}
+
+	// S3 doesn't have the concept of empty directories, so it'll return path not found if there are no objects
+	if objectCount == 0 {
+		return storagedriver.PathNotFoundError{Path: from}
+	}
+
+	return nil
+}
+
+// WalkFiles traverses a filesystem defined within driver, starting
+// from the given path, calling f on each file
+func (d *driver) WalkFiles(ctx context.Context, from string, f storagedriver.WalkFn) error {
```

Comment:
Should we update the [walk_test](https://github.com/digitalocean/docker-distribution/blob/master/registry/storage/driver/walk_test.go) for this slightly different logic?

Reply:
There aren't even tests for `Walk` really. I don't quite want to get into the work of adding tests for them but could if it's worth the effort

Reply:
Yeah good point. But still worth the consideration, especially with regards to preventing regressions in functionality. Lack of proper testing for current functionality is never a reason for not adding tests :wink:

Reply:
Turns out my fallback method was actually incorrect. Added a bunch of unit tests to check the behavior more explicitly.

Reply:
I'd like to add S3 tests as well but that's way more involved, I don't want to spent too too long on this... 😭